aarch64 Interrupt timer fix and NS-EL1 support

MrVan opened the pull request:
This is to unmask aarch64 generic timer interrupt and make it could work in NS-EL1.

For the SPI part, I am not sure it might break your current settings. I kick zephyr in NS-EL1 state, not in EL3.

MrVan committed:
drivers: interrupt_controller: make GICv3 work for NS-EL1

Configure SPIs/SGI1 as NS interrupt
Enable ARE_NS, G1NS

Signed-off-by: Peng Fan <peng.fan@nxp.com>

carlocaione commented on include/arch/arm/aarch64/timer.h:
```diff
@@ -20,7 +20,8 @@ extern "C" {
 
-#define CNTV_CTL_ENABLE		((1) << 0)
+#define CNTV_CTL_ENABLE		BIT(0)
+#define CNTV_CTL_IT_MASK	BIT(1)
```
Missing tab

MrVan replied:
hmm. VIM shows they are in same col. I'll check more.

carlocaione commented:
@sandeepbrcm @Abhishek-brcm can you check whether this works on your hardware?

carlocaione commented on drivers/interrupt_controller/intc_gicv3.c:
```diff
@@ -249,14 +253,18 @@ static void gicv3_dist_init(void)
 			    ICPENDR(base, idx));
 		/* All SPIs are G1S and owned by Zephyr */
-		sys_write32(0, IGROUPR(base, idx));
+		sys_write32(~0, IGROUPR(base, idx));
```
The comment needs to be updated

carlocaione replied:
What happens if Zephyr is running in NS mode?

sandeepbrcm replied:
The TODOs in this driver roughly marks where we need to update to support NS state. 
Apart from the GICD we need CPU interface routed accordingly and conditionally. We dont want to drop support for 
running Zephyr in Secure state as a standalone RTOS ( start EL3 ->run EL1S).
ref: https://github.com/zephyrproject-rtos/zephyr/blob/0277b4f071818e836ac4ed1631abfa425a9dd7e9/drivers/interrupt_controller/intc_gicv3.c#L183-L189

MrVan replied:
ok. Then need to introduce an extra Kconfig option for zephyr in secure and non-secure stage.

carlocaione commented on drivers/interrupt_controller/intc_gicv3.c:
```diff
@@ -249,14 +253,18 @@ static void gicv3_dist_init(void)
 	/* wait for rwp on GICD */
 	gic_wait_rwp(GIC_SPI_INT_BASE);
 
+	/* Enable distributor with ARE*/
```
nit: missing space after `ARE`

carlocaione commented:
Tagging @sandeepbrcm because he wrote the driver.

@MrVan I have a question about this whole patchset. It looks like you want to use the GIC driver from EL1NS but to me it looks like the whole driver is written with EL1S in mind. Take for example `GICD_CTLR`, in this case the bitfields are different when access is secure vs non-secure (for example `ARE_NS` is `BIT(5)` when access is secure but `BIT(4)` when access is non-secure).

sandeepbrcm commented:
> Tagging @sandeepbrcm because he wrote the driver.
> 
> @MrVan I have a question about this whole patchset. It looks like you want to use the GIC driver from EL1NS but to me it looks like the whole driver is written with EL1S in mind. Take for example `GICD_CTLR`, in this case the bitfields are different when access is secure vs non-secure (for example `ARE_NS` is `BIT(5)` when access is secure but `BIT(4)` when access is non-secure).

Ditto. Thanks.  @MrVan Your attempt to add non-secure EL1 support is almost right and valid. However it needs a little more consideration. Just wondering if you really want to run Zephyr in Non-secure state (EL1-NS) or EL1 Secure is acceptable. 
ie: Are you going to use TZ partitioning.  Same question for you @carlocaione .

carlocaione commented:
> Ditto. Thanks. @MrVan Your attempt to add non-secure EL1 support is almost right and valid. However it needs a little more consideration. Just wondering if you really want to run Zephyr in Non-secure state (EL1-NS) or EL1 Secure is acceptable.
> ie: Are you going to use TZ partitioning. Same question for you @carlocaione .

Yes, running Zephyr in EL1NS is definitely something we want in the future. Not high priority for me at the moment.

sandeepbrcm commented on drivers/interrupt_controller/intc_gicv3.c:
```diff
@@ -249,14 +253,18 @@ static void gicv3_dist_init(void)
 	gic_wait_rwp(GIC_SPI_INT_BASE);
 
+	/* Enable distributor with ARE*/
+	sys_write32(BIT(GICD_CTRL_ARE_NS) | BIT(GICD_CTLR_ENABLE_G1NS) |
+		    BIT(GICD_CTLR_ENABLE_G0), GICD_CTLR);
+
```
Prevailing guidance on GICv3 hardware integration is to tie ARE  'HIGH' as legacy mode is deprecated feature.
But no harm in configuring explicitly in case some SOC implementation still provides ARE configuration.

sandeepbrcm replied:
Instead of updating GICD_CTLR here you can update at line 281  G1NS/G1S based on current security state.

sandeepbrcm replied:
'GICD_CTLR_ENABLE_G0'  can not take that to EL1NS.

MrVan commented:
> > Tagging @sandeepbrcm because he wrote the driver.
> > @MrVan I have a question about this whole patchset. It looks like you want to use the GIC driver from EL1NS but to me it looks like the whole driver is written with EL1S in mind. Take for example `GICD_CTLR`, in this case the bitfields are different when access is secure vs non-secure (for example `ARE_NS` is `BIT(5)` when access is secure but `BIT(4)` when access is non-secure).
> 
> Ditto. Thanks. @MrVan Your attempt to add non-secure EL1 support is almost right and valid. However it needs a little more consideration. Just wondering if you really want to run Zephyr in Non-secure state (EL1-NS) or EL1 Secure is acceptable.
> ie: Are you going to use TZ partitioning. Same question for you @carlocaione .

No. I need run zephyr on top of a NS hypervisor, with zephyr inside a VM, linux inside a VM. Previously I thought little kernel might be a better choice which has better ARM64 SMP support, but it really not have a good community, so I choose zephyr which also means more work :)

sandeepbrcm commented:
> > > Tagging @sandeepbrcm because he wrote the driver.
> > > @MrVan I have a question about this whole patchset. It looks like you want to use the GIC driver from EL1NS but to me it looks like the whole driver is written with EL1S in mind. Take for example `GICD_CTLR`, in this case the bitfields are different when access is secure vs non-secure (for example `ARE_NS` is `BIT(5)` when access is secure but `BIT(4)` when access is non-secure).
> > 
> > 
> > Ditto. Thanks. @MrVan Your attempt to add non-secure EL1 support is almost right and valid. However it needs a little more consideration. Just wondering if you really want to run Zephyr in Non-secure state (EL1-NS) or EL1 Secure is acceptable.
> > ie: Are you going to use TZ partitioning. Same question for you @carlocaione .
> 
> No. I need run zephyr on top of a NS hypervisor, with zephyr inside a VM, linux inside a VM. Previously I thought little kernel might be a better choice which has better ARM64 SMP support, but it really not have a good community, so I choose zephyr which also means more work :)

Yes. But no unknowns from GIC driver perspective for SMP or to run in EL1_NS. I think the changes I mentioned should get you going with NS state. I had deliberately chosen the PMR and priorities in that range so that it works for both S/NS. And the driver was tested with EL1_NS in a Broadcom version of aarch64 port long back.

And yes a config is better throughout Zephyr as there will be other places eg: the MMU default mapping have to drop the secure access attribute and may be some other places.

For SMP GIC driver redistributors are per cpu and with right addresses it should work.  SGI generation should work without change.

lorc committed:
arch: arm: timer: mask interrupt in ISR

As timer interrupt is level triggered, we need to mask it before leaving
ISR or it will be delivered again.

Also, Xen automatically masks timer interrupt when it injects IRQ to
a guest, so we need to unmask it again, when setting new timeout.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>
Signed-off-by: Peng Fan <peng.fan@nxp.com>

MrVan committed:
arm: aarch64: Kconfig: introduce ARMV8_A_NS

Add new Kconfig entry ARMV8_A_NS for zephyr runs in normal world usage.

Signed-off-by: Peng Fan <peng.fan@nxp.com>

MrVan commented:
Updated with more patches to make qemu cortex a53 could run in NS world. Please help reivew, thanks.

MrVan committed:
arch: arm64: mmu: support using MT_NS attribute

According to CONFIG_ARMV8_A_NS, using MT_SECURE or MT_NS, to simplify
code change, use MT_DEFAULT_SECURE_STATE instead

Signed-off-by: Peng Fan <peng.fan@nxp.com>

MrVan committed:
arm: qemu_cortex_a53: support running in NS world

Support qemu_cortex_a53 runs in Normal World

Signed-off-by: Peng Fan <peng.fan@nxp.com>

carlocaione commented on arch/arm/core/aarch64/Kconfig:
```diff
@@ -58,6 +58,12 @@ config IPM_CONSOLE_STACK_SIZE
 if CPU_CORTEX_A
 
+config ARMV8_A_NS
+	bool "ARMv8-A Normal World(None Secure world of Trustzonel)"
```
typos: `ARMv8-A Normal World (Non-Secure world of Trustzone)`

MrVan replied:
fix in next version

carlocaione commented on boards/arm/qemu_cortex_a53/board.cmake:
```diff
@@ -5,9 +5,17 @@ set(EMU_PLATFORM qemu)
 set(QEMU_ARCH aarch64)
 
 set(QEMU_CPU_TYPE_${ARCH} cortex-a53)
+if("${CONFIG_ARMV8_A_NS}" STREQUAL "y")
+set(QEMU_FLAGS_${ARCH}
+  -cpu ${QEMU_CPU_TYPE_${ARCH}}
+  -nographic
+  -machine virt,gic-version=3
+  )
+else()
```
No need to use `STREQUAL` with `CONFIG_*` boolean symbols. You can just use `if(CONFIG_ARMV8_A_NS)...`.

Also instead of duplicating all the options can you just introduce `QEMU_EXTRA_FLAGS` and use that to differentiate between the two cases?

carlocaione commented on drivers/interrupt_controller/intc_gicv3.c:
```diff
@@ -248,8 +254,10 @@ static void gicv3_dist_init(void)
 			    ICPENDR(base, idx));
-		/* All SPIs are G1S and owned by Zephyr */
-		sys_write32(0, IGROUPR(base, idx));
+		if (IS_ENABLED(CONFIG_ARMV8_A_NS))
```
`CONFIG_ARMV8_A_NS` is a boolean and is guarding a pretty trivial `sys_write32` so why don't you just use an `ifdef`?

MrVan replied:
fix in next version.

carlocaione commented on include/arch/arm/aarch64/timer.h:
```diff
@@ -20,7 +20,8 @@ extern "C" {
 
-#define CNTV_CTL_ENABLE		((1) << 0)
+#define CNTV_CTL_ENABLE		BIT(0)
+#define CNTV_CTL_IT_MASK	BIT(1)
```
If you want to keep the `_MASK` suffix please fix `CNTV_CTL_ENABLE` to keep it consistent. Otherwise just remove it.

MrVan replied:
You mean tab aligned? In VIM, the two lines are aligned with linux coding style, not sure why github shows difference.

carlocaione replied:
Nope, I mean that you use the `_MASK` suffix for `CNTV_CTL_IT_MASK` but `CNTV_CTL_ENABLE` has no suffix at all. So either you use just `CNTV_CTL_IT` or you fix `CNTV_CTL_ENABLE` to `CNTV_CTL_ENABLE_MASK` for the sake of consistency.

carlocaione commented on include/arch/arm/aarch64/arm_mmu.h:
```diff
@@ -67,6 +67,12 @@
 #define MT_P_RX_U_RX		(MT_RO | MT_RW_AP_ELx | MT_P_EXECUTE | MT_U_EXECUTE)
 #define MT_P_RX_U_NA		(MT_RO | MT_RW_AP_EL_HIGHER  | MT_P_EXECUTE | MT_U_EXECUTE_NEVER)
 
+#if IS_ENABLED(CONFIG_ARMV8_A_NS)
+#define MT_SECURE_ATTR MT_NS
+#else
+#define MT_SECURE_ATTR MT_SECURE
+#endif
```
`MT_SECURE_ATTR` is a bit too opaque to me, what about `MT_DEFAULT_SECURE_STATE` or something like that to point out that we are using the "default" state Zephyr was compiled for?

Also, please use `ifdef` for consistency.

MrVan replied:
fix in next version.

MrVan commented:
All, updated and rebased, please help review.

carlocaione commented:
@MrVan kinda curious how you are testing this on QEMU. I was expecting to see the `NS` bit of the `SCR_EL3` set somewhere. 

I quickly tried to enter in `NS` setting the `SCR_EL3_NS` bit in `SCR_EL3` but this is not working.

carlocaione commented on include/arch/arm/aarch64/timer.h:
```diff
@@ -36,10 +37,13 @@ static ALWAYS_INLINE void arm_arch_timer_enable(unsigned char enable)
-		cntv_ctl &= ~CNTV_CTL_ENABLE;
+	if (enable) {
+		cntv_ctl |= CNTV_CTL_ENABLE_MASK;
+		cntv_ctl &= ~CNTV_CTL_IT_MASK;
```
Can you evaluate the fix proposed in https://github.com/zephyrproject-rtos/zephyr/pull/30164/commits/be5244b61d33e481320faa77a8a00f3886ab6ca4 ?

MrVan replied:
ok, that fix seems better.

MrVan commented:
> @MrVan kinda curious how you are testing this on QEMU. I was expecting to see the `NS` bit of the `SCR_EL3` set somewhere.
> 
> I quickly tried to enter in `NS` setting the `SCR_EL3_NS` bit in `SCR_EL3` but this is not working.

@carlocaione , I just use `west build -t run` to do the test. If 'secure=on' not passed to qemu, SCR_EL3 is not accessible. On my real board, I do have ATF do that stuff, but with qemu, I no need ATF, unless I wanna optee runs in secure world.

carlocaione commented on boards/arm/qemu_cortex_a53/board.cmake:
```diff
@@ -5,9 +5,16 @@ set(EMU_PLATFORM qemu)
 set(QEMU_CPU_TYPE_${ARCH} cortex-a53)
+
+if(CONFIG_ARMV8_A_NS)
+set(QEMU_MACH virt,virtualization=on,gic-version=3)
```
This is confusing because `CONFIG_ARMV8_A_NS` has nothing to do with you change here. You removed `secure=on` and added `virtualization=on`, this means that QEMU is now starting from EL2. 

- What's the relation with `NS/S` world?

My expectation here is that you can have QEMU running Zephyr in `NS` world but this is not the case because when you enable `CONFIG_ARMV8_A_NS` all the QEMU tests stop working (because starting from EL2 is not supported).

- What is this change for considering that `CONFIG_ARMV8_A_NS` is actually breaking the QEMU emulation?

carlocaione commented on boards/arm/qemu_cortex_a53/board.cmake:
```diff
@@ -5,9 +5,16 @@ set(EMU_PLATFORM qemu)
 set(QEMU_CPU_TYPE_${ARCH} cortex-a53)
+
+if(CONFIG_ARMV8_A_NS)
+set(QEMU_MACH virt,virtualization=on,gic-version=3)
```
This is confusing because CONFIG_ARMV8_A_NS has nothing to do with you change here. You removed secure=on and added virtualization=on, this means that QEMU is now starting from EL2.

- What's the relation with NS/S world?

My expectation here is that you can have QEMU running Zephyr in NS world but this is not the case because when you enable CONFIG_ARMV8_A_NS all the QEMU tests stop working (because starting from EL2 is not supported).

- What is this change for considering that CONFIG_ARMV8_A_NS is actually breaking the QEMU emulation?

MrVan replied:
I have dropped virtualization=on, this is added by mistake when I support zephyr boot from EL2.

MrVan replied:
I not added a new defconfig now, just provide this option to let people could use it. The final goal as I said I will make zephyr run on top of jailhouse hypervisor, and make zephyr boot from NS-EL2 to make it could be root cell.

carlocaione commented:
> @carlocaione , I just use `west build -t run` to do the test. If 'secure=on' not passed to qemu, SCR_EL3 is not accessible. On my real board, I do have ATF do that stuff, but with qemu, I no need ATF, unless I wanna optee runs in secure world.

My question is: *can you run Zephyr in NS mode without any secure monitor under QEMU?*
- (yes) If we can then I expect that to be tested and working because right now enabling `CONFIG_ARMV8_A_NS` is actually breaking QEMU emulation
- (no) If we can not, then I do not understand your changed to the QEMU parameters and also you should explicitly clarify that somewhere (documentation?)

MrVan commented:
> > @carlocaione , I just use `west build -t run` to do the test. If 'secure=on' not passed to qemu, SCR_EL3 is not accessible. On my real board, I do have ATF do that stuff, but with qemu, I no need ATF, unless I wanna optee runs in secure world.
> 
> My question is: _can you run Zephyr in NS mode without any secure monitor under QEMU?_

Yes

> 
> * (yes) If we can then I expect that to be tested and working because right now enabling `CONFIG_ARMV8_A_NS` is actually breaking QEMU emulation

I forgot to remove virtualization args. I'll update the patch.
-set(QEMU_MACH virt,virtualization=on,gic-version=3)
+set(QEMU_MACH virt,gic-version=3)

> * (no) If we can not, then I do not understand your changed to the QEMU parameters and also you should explicitly clarify that somewhere (documentation?)

Just drop secure=on is enough.

MrVan commented:
Updated. Please help review! Thanks.

carlocaione commented:
I think we are not understanding each other on what `CONFIG_ARMV8_A_NS` is meant to mean :)

> > My question is: _can you run Zephyr in NS mode without any secure monitor under QEMU?_
> 
> Yes

Ok, how? (see later)

> > * (yes) If we can then I expect that to be tested and working because right now enabling `CONFIG_ARMV8_A_NS` is actually breaking QEMU emulation
> 
> I forgot to remove virtualization args. I'll update the patch.
> -set(QEMU_MACH virt,virtualization=on,gic-version=3)
> +set(QEMU_MACH virt,gic-version=3)

This is not solving anything. With this patch when you enable `CONFIG_ARMV8_A_NS` you are basically now running Zephyr under QEMU starting from EL1. How this is related to the fact that you want it running in NS mode?

Or there is a misunderstanding about what `CONFIG_ARMV8_A_NS` actually means?

MrVan commented:
> I think we are not understanding each other on what `CONFIG_ARMV8_A_NS` is meant to mean :)
> 
> 
> 
> > > My question is: _can you run Zephyr in NS mode without any secure monitor under QEMU?_
> 
> > 
> 
> > Yes
> 
> 
> 
> Ok, how? (see later)
> 
> 
> 
> > > * (yes) If we can then I expect that to be tested and working because right now enabling `CONFIG_ARMV8_A_NS` is actually breaking QEMU emulation
> 
> > 
> 
> > I forgot to remove virtualization args. I'll update the patch.
> 
> > -set(QEMU_MACH virt,virtualization=on,gic-version=3)
> 
> > +set(QEMU_MACH virt,gic-version=3)
> 
> 
> 
> This is not solving anything. With this patch when you enable `CONFIG_ARMV8_A_NS` you are basically now running Zephyr under QEMU starting from EL1. How this is related to the fact that you want it running in NS mode?
> 
> 

secure=on is not passed to qemu, so qemu runs in NS-EL1

> 
> Or there is a misunderstanding about what `CONFIG_ARMV8_A_NS` actually means?
> 
> 
> 
> 

I mean let zephyr runs in non secure world when it set to Y.

> 
>

carlocaione commented:
> secure=on is not passed to qemu, so qemu runs in NS-EL1

Correct. Now, this is a bit confusing because without `CONFIG_ARMV8_A_NS` you are starting from EL3 and dropping into S-EL1, with `CONFIG_ARMV8_A_NS` you are directly starting from NS-EL1. I'm not exactly sure if you can start from EL3 and dropping into NS-EL1 with QEMU (because this is what I was expecting to see when enabling this option).

Anyway, please in the help to `CONFIG_ARMV8_A_NS` clarify this point, that is when you enable the option you start directly from NS-EL1.

carlocaione commented:
@MrVan ok, I think I have cleared my doubts. Before merging this could you please cherry-pick be5244b61d33e481320faa77a8a00f3886ab6ca4 instead of your fix?

MrVan commented:
> > secure=on is not passed to qemu, so qemu runs in NS-EL1
> 
> 
> 
> Correct. Now, this is a bit confusing because without `CONFIG_ARMV8_A_NS` you are starting from EL3 and dropping into S-EL1, with `CONFIG_ARMV8_A_NS` you are directly starting from NS-EL1. I'm not exactly sure if you can start from EL3 and dropping into NS-EL1 with QEMU (because this is what I was expecting to see when enabling this option).
> 

Starting from EL3 and drop to NS-EL1 means       I will not able to use Arm trusted firmware on real board

> 
> 
> ~Anyway, please in the help to `CONFIG_ARMV8_A_NS` clarify this point, that is when you enable the option you start directly from NS-EL1.~ Probably not a good idea either because this is something related to QEMU.

ok, will add the help for the option

carlocaione commented on arch/arm/core/aarch64/Kconfig:
```diff
@@ -58,6 +58,12 @@ config IPM_CONSOLE_STACK_SIZE
+config ARMV8_A_NS
+	bool "ARMv8-A Normal World (Non-Secure world of Trustzone)"
+	help
+	  This option signifies the use of an ARMv8-A processor
```
Wouldn't be better to say that is Zephyr that is running in Trustzone Normal World?

MrVan replied:
Done

carlocaione commented:
> ok, will add the help for the option

No need because that is only true if we are running Zephyr on QEMU and that is not the case if you are running it on real hardware.

MrVan commented:
> @MrVan ok, I think I have cleared my doubts. Before merging this could you please cherry-pick [be5244b](https://github.com/zephyrproject-rtos/zephyr/commit/be5244b61d33e481320faa77a8a00f3886ab6ca4) instead of your fix?

Done

MrVan commented:
Updated. Please help review. Thanks

carlocaione commented on arch/arm/core/aarch64/Kconfig:
```diff
@@ -58,6 +58,12 @@ config IPM_CONSOLE_STACK_SIZE
+config ARMV8_A_NS
+	bool "ARMv8-A Normal World (Non-Secure world of Trustzone)"
+	help
+	  This option signifies Zephyr that is running in Trustzone
```
s/signifies Zephyr that/signifies that zephyr/

MrVan replied:
Done

carlocaione commented:
besides the nit about the help LGTM. Thank you @MrVan for putting up with my requests.

MrVan commented:
@sandeepbrcm @andrewboie @Abhishek-brcm Please help review. Thanks.

sandeepbrcm commented on arch/arm/core/aarch64/Kconfig:
```diff
@@ -58,6 +58,12 @@ config IPM_CONSOLE_STACK_SIZE
+	bool "ARMv8-A Normal World (Non-Secure world of Trustzone)"
+	help
+	  This option signifies that Zephyr is running in Trustzone
+	  Normal World.
```
Should we make it explicit Zephyr is entered in Trustzone Non-secure state.
Overall the PR looks good.

MrVan replied:
I am fine with default NS or S-EL3, just default y or select it. But it may break others which already use zephyr in EL3. So I would not change it to default currently. Until we could migrate all arm64 boards to NS, we could make it default. @carlocaione Any comments?

sandeepbrcm replied:
I just meant to update the config help  "This option signifies that Zephyr is running in Trustzone
	  Normal World."   to  "This option signifies that Zephyr is **entered** in Trustzone
	  Non-Secure state".

Default option is better to leave at EL3 ie "n" and change later depending on how all platforms use it.

MrVan replied:
@sandeepbrcm  Updated. Please help approve this PR.

MrVan commented:
@MaureenHelm Please help merge this PR. Thanks.

stephanosio commented on drivers/interrupt_controller/intc_gicv3.c:
```diff
@@ -180,12 +185,11 @@ static void gicv3_cpuif_init(void)
-	/* Configure all SGIs/PPIs as G1S.
-	 * TODO: G1S or G1NS dependending on Zephyr is run in EL1S
-	 * or EL1NS respectively.
+	/* Configure all SGIs/PPIs as G1S or G1NS dependending on Zephyr
```
Typo
```suggestion
	/* Configure all SGIs/PPIs as G1S or G1NS depending on Zephyr
```

stephanosio commented on include/arch/arm/aarch64/timer.h:
```diff
@@ -45,6 +46,22 @@ static ALWAYS_INLINE void arm_arch_timer_enable(unsigned char enable)
+	__asm__ volatile("mrs %0, cntv_ctl_el0\n\t"
+			 : "=r" (cntv_ctl) :  : "memory");
+
+	if (mask)
+		cntv_ctl |= CNTV_CTL_IMASK;
+	else
+		cntv_ctl &= ~CNTV_CTL_IMASK;
```
Unless there has been a change in the coding guidelines (please do let me know if there has; I've been away for a while), these should have curly braces.

```suggestion
	if (mask) {
		cntv_ctl |= CNTV_CTL_IMASK;
	} else {
		cntv_ctl &= ~CNTV_CTL_IMASK;
	}
```

carlocaione replied:
@stephanosio glad you are back :)

stephanosio commented on include/arch/arm/aarch64/timer.h:
```diff
@@ -45,6 +46,22 @@ static ALWAYS_INLINE void arm_arch_timer_enable(unsigned char enable)
 			 : : "r" (cntv_ctl) : "memory");
 }
 
+static ALWAYS_INLINE void arm_arch_timer_irq_mask(unsigned char mask)
```
Boolean-valued params should use the `bool` type
```suggestion
static ALWAYS_INLINE void arm_arch_timer_irq_mask(bool mask)
```

stephanosio replied:
It is often desirable to name the functions that set/get values accordingly.

```suggestion
static ALWAYS_INLINE void arm_arch_timer_set_irq_mask(unsigned char mask)
```

stephanosio commented on arch/arm/core/aarch64/Kconfig:
```diff
@@ -58,6 +58,12 @@ config IPM_CONSOLE_STACK_SIZE
 
 if CPU_CORTEX_A
 
+config ARMV8_A_NS
+	bool "ARMv8-A Normal World (Non-Secure world of Trustzone)"
+	help
+	  This option signifies that Zephyr is entered in Trustzone
+	  Non-Secure state
```
Nit: the official capitalisation is `TrustZone`.
```suggestion
config ARMV8_A_NS
	bool "ARMv8-A Normal World (Non-Secure world of TrustZone)"
	help
	  This option signifies that Zephyr is entered in TrustZone
	  Non-Secure state
```

stephanosio reviewed (commented):
Just some generic comments.

nashif commented on include/arch/arm/aarch64/timer.h:
```diff
@@ -45,6 +46,22 @@ static ALWAYS_INLINE void arm_arch_timer_enable(unsigned char enable)
+	__asm__ volatile("mrs %0, cntv_ctl_el0\n\t"
+			 : "=r" (cntv_ctl) :  : "memory");
+
+	if (mask)
```
please add curly braces.

MrVan commented:
@stephanosio @nashif Updated & rebased. Please check. Thanks.
